Include/internal/pycore_frame.h

```diff
@@ -258,6 +260,35 @@ _PyThreadState_PushFrame(PyThreadState *tstate, size_t size);
 
 void _PyThreadState_PopFrame(PyThreadState *tstate, _PyInterpreterFrame *frame);
 
+
+/* Converts frame for tier 2.
+ * Adds stack space at the end of the current frame for Tier 2 execution.
+ * The frame that is being expanded MUST be the current executing frame, and
+ * it must be at the top of the datastack.
+ * */
+static inline int
+_PyFrame_ConvertToTier2(PyThreadState *tstate, _PyInterpreterFrame *frame,
+                        int localsplus_grow)
+{
+    assert(localsplus_grow > 0);
+    // Already grown previously
+    if (frame->tier2_extra_size >= localsplus_grow) {
+        return 0;
+    }
+    if (frame->owner != FRAME_OWNED_BY_THREAD) {
+        return 1;
+    }
+    if (!_PyThreadState_HasStackSpace(tstate, localsplus_grow)) {
```

Comment:
I believe there's a bug in this code that can be solved by inserting
```c
localsplus_grow -= frame->tier2_extra_size;
```
here.

My reasoning is that `grow` is how many entries we require (at least) to exist from the end of the "official" frame size (which points to the end of the reserved stack space). So if we request 5 slots and we've already grown 3 slots, we should only need to grow another 2 slots, but your code here would add 5 slots in that case. That never crashes, but it over-allocates. The scenario is very unlikely, but not impossible (as you explained in the other discussion thread). Eventually everything always settles, because `tier2_extra_size` accurately keeps track of the over-allocation. Hence my proposal to adjust `grow` at this point, once, so the remaining invariants remain intact.